# Import Libraries

In [2]:
# importing libraries
import pandas as pd
import numpy as nm
import os

# Import Data

In [27]:
#importing datasets products and prepared orders
path = r'/Users/mohammedRusso/Documents/DATA_ANALYST_COURSE/DATA_Immersion_Submissions/Assingment 4/01-2025 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [29]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [21]:
df_ords['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    600905
1    587478
2    467260
5    453368
6    448761
3    436972
4    426339
Name: count, dtype: int64

In [23]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [45]:
df_ords = df_ords.drop(columns=['Unnamed: 0'])

In [47]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [49]:
#create a Dataframe
df_test = pd.DataFrame()

In [53]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [55]:
df_test

mix
0     a
1     b
2     1
3  True

In [63]:
#check for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [65]:
df_test['mix'] = df_test['mix'].astype('str')

In [67]:
df_test

mix
0     a
1     b
2     1
3  True

In [75]:
df_test.info

<bound method DataFrame.info of     mix
0     a
1     b
2     1
3  True>

In [83]:
df_a = df_test['mix']== 'a'

In [85]:
df_a

0     True
1    False
2    False
3    False
Name: mix, dtype: bool

In [95]:
#find missing values in every column in the table
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [97]:
# find null values for a specific column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [93]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [ ]:
#impute column with mean value
df['column with missings'].fillna(mean value, inplace=True)
#impute column with median value
df['column with missings'].fillna(median value, inplace=True)

In [99]:
df_prods.shape

(49693, 5)

In [103]:
#remove all null value from specified column
df_prods_clean = df_prods[df_prods['product_name'].isnull()== False]

In [105]:
df_prods_clean.shape

(49677, 5)

In [ ]:
#drop all missing values from whole table --DO NOT EXECITE--
df_prods.dropna(inplace = True)

In [ ]:
#drop only the NaNs from a particular column -- DO NOT EXECUTE--
df_prods.dropna(subset = [‘product_name’], inplace = True)

In [107]:
# Look for duplicate within the dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [109]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [117]:
#drop duplicate roew from table
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [119]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting changes made to Products

In [121]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 4.5 Task

#### STEP 2

In [154]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

##### Both min and max of the 'order_number' column is 1 indicating there are many duplicate values. 
'order_day of the week': Has min value of 0 and max value of 6 meaning the seven numbers (0-6) represent days of the week.

'order_hour_of_day': Has a min value of 0 and max value of 23 meaning the twenty-four numbers (0-23) represent 24 hrs of the day.

There are no unuasul or outlier values present in the orders dataframe

#### STEP 3 Check for mixed-type data in your df_ords dataframe.

In [167]:
#check for mixed data types 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [169]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


##### There is no mixed-type data found in any of the columns/variables in the df_ords dataframe.

#### STEP 5 Run a check for missing values in your df_ords dataframe.

In [188]:
#checking for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### The 'days_since_prior_order' is the only column is missing values (206209). Likely causes are either data corruption or they were never recorded

#### STEP 6 Address the missing values using an appropriate method.

In [194]:
#Addressing missing values by creating a subset
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [192]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

##### The missing values coincides with the number of unique 'user_id' so there is no need to address these values

#### STEP 7 & 8 Run a check for duplicate values in your df_ords data.

In [199]:
# Checking for duplicate values 
df_ords_dups = df_ords[df_ords.duplicated()]

In [201]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

##### There are no duplicate values in the orders dataframe. 

#### STEP 9 Export your final, cleaned df_prods and df_ords data

In [208]:
# Droping unnamed column
df_ords_cln = df_ords.drop(columns=['Unnamed: 0'], axis=1)

In [210]:
df_ords_cln

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [212]:
df_ords_cln.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))